File: 2nd Year/2nd Semester/MPP/PROIECT-C#/Lab1/Lab1/Repository/ParticipantRepository.cs
using System;
using System.Collections.Generic;
using System.Configuration;
using System.Data.SQLite;
using Lab1.Domain;
using log4net;

namespace Lab1.Repository {
    public class ParticipantRepository : IParticipantRepository {
        private static readonly ILog logger = LogManager.GetLogger(typeof(ParticipantRepository));
        public readonly string connectionString;

        public ParticipantRepository() {
            connectionString = ConfigurationManager.ConnectionStrings["DefaultConnection"].ConnectionString;
        }

        public void Delete(int id) {
            logger.Debug("Enter deleting Participant with id: " + id);
            string sql = "DELETE FROM Participant WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", id);
                        command.ExecuteNonQuery();
                    }
                    connection.Close();
                }
                logger.Debug("Exit deleting Participant with id: " + id);
            }
            catch (SQLiteException ex) {
                logger.Error("Error deleting Participant with id: " + id, ex);
            }
        }

        public List<Participant> FindAll() {
            logger.Debug("Enter FindAll Participanti: ");
            List<Participant> participanti = new List<Participant>();
            string sql = "SELECT * FROM Participant";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        using (SQLiteDataReader reader = command.ExecuteReader()) {
                            while (reader.Read()) {
                                int id = reader.GetInt32(0);
                                string firstName = reader.GetString(1);
                                string lastName = reader.GetString(2);
                                Participant participant = new Participant(id, firstName, lastName);
                                participanti.Add(participant);
                                logger.Debug("Found Participant: " + participant);
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex) {
                logger.Error("Error FindAll Participanti", ex);
            }
            return participanti;
        }

        public Participant FindOne(int id) {
            logger.Debug("Enter FindOne Participant with id: " + id);
            string sql = "SELECT * FROM Participant WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", id);
                        using (SQLiteDataReader reader = command.ExecuteReader()) {
                            if (reader.Read()) {
                                string firstName = reader.GetString(1);
                                string lastName = reader.GetString(2);
                                Participant participant = new Participant(id, firstName, lastName);
                                logger.Debug("Found Participant: " + participant);
                                return participant;
                            }
                        }
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex) {
                logger.Error("Error FindOne Participant with id: " + id, ex);
            }
            return null;
        }

        public void Save(Participant entity) {
            logger.Debug("Enter Save Participant: " + entity);
            string sql = "INSERT INTO Participant (first_name, last_name) VALUES (@firstName, @lastName)";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@firstName", entity.FirstName);
                        command.Parameters.AddWithValue("@lastName", entity.LastName);

                        int affectedRows = command.ExecuteNonQuery();
                        if (affectedRows > 0) {
                            command.CommandText = "SELECT last_insert_rowid()";
                            entity.Id = Convert.ToInt32(command.ExecuteScalar());
                        }
                        logger.Debug("Saved Participant: " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex) {
                logger.Error("Error saving Participant: " + entity, ex);
            }
        }

        public void Update(Participant entity) {
            logger.Debug("Enter Update Participant: " + entity);
            string sql = "UPDATE Participant SET first_name=@firstName, last_name=@lastName WHERE id=@id";
            try {
                using (SQLiteConnection connection = new SQLiteConnection(connectionString)) {
                    connection.Open();
                    using (SQLiteCommand command = new SQLiteCommand(sql, connection)) {
                        command.Parameters.AddWithValue("@id", entity.Id);
                        command.Parameters.AddWithValue("@firstName", entity.FirstName);
                        command.Parameters.AddWithValue("@lastName", entity.LastName);
                        command.ExecuteNonQuery();
                        logger.Debug("Updated Participant: " + entity);
                    }
                    connection.Close();
                }
            }
            catch (SQLiteException ex) {
                logger.Error("Error updating Participant: " + entity, ex);
            }
        }
    }
}

File: 2nd Year/2nd Semester/MPP/Triatlon-Java/src/main/java/com/example/laborator/Repository/ParticipantRepository.java
package com.example.laborator.Repository;

import com.example.laborator.Domain.Participant;
import com.example.laborator.Utils.JdbcUtils;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Properties;

public class ParticipantRepository implements IParticipantRepository {
    private static final Logger logger = LogManager.getLogger();
    private final JdbcUtils jdbcUtils;

    public ParticipantRepository(Properties properties) {
        logger.info("Initializing ParticipantRepository with proprieties: {}", properties);
        jdbcUtils = new JdbcUtils(properties);
    }

    @Override
    public Participant findOne(Integer integer) {
        logger.traceEntry("finding Participant with id {}", integer);
        String sql = "SELECT * FROM Participant WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);

            try (ResultSet rs = preparedStatement.executeQuery()) {
                if (rs.next()) {
                    return new Participant(
                            rs.getInt("id"),
                            rs.getString("first_name"),
                            rs.getString("last_name")
                    );
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        return null;
    }

    @Override
    public Iterable<Participant> findAll() {
        List<Participant> participants = new ArrayList<>();
        logger.traceEntry("finding all Participants");
        String sql = "SELECT * FROM Participant";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql);
             ResultSet resultSet = preparedStatement.executeQuery()) {

            while (resultSet.next()) {
                participants.add(new Participant(
                        resultSet.getInt("id"),
                        resultSet.getString("first_name"),
                        resultSet.getString("last_name")
                ));
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
        return participants;
    }

    @Override
    public void save(Participant entity) {
        logger.traceEntry("saving Participant {}", entity);
        String sql = "INSERT INTO Participant (first_name, last_name) VALUES (?, ?)";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql, PreparedStatement.RETURN_GENERATED_KEYS)) {

            preparedStatement.setString(1, entity.getFirst_name());
            preparedStatement.setString(2, entity.getLast_name());

            int affectedRows = preparedStatement.executeUpdate();
            if (affectedRows > 0) {
                try (ResultSet generatedKeys = preparedStatement.getGeneratedKeys()) {
                    if (generatedKeys.next()) {
                        entity.setId(generatedKeys.getInt(1));
                    }
                }
            }
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void delete(Integer integer) {
        logger.traceEntry("deleting Participant with id {}", integer);
        String sql = "DELETE FROM Participant WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setInt(1, integer);
            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }

    @Override
    public void update(Participant entity) {
        logger.traceEntry("updating Participant {}", entity);
        String sql = "UPDATE Participant SET first_name = ?, last_name = ? WHERE id = ?";

        try (Connection connection = jdbcUtils.getConnection();
             PreparedStatement preparedStatement = connection.prepareStatement(sql)) {

            preparedStatement.setString(1, entity.getFirst_name());
            preparedStatement.setString(2, entity.getLast_name());
            preparedStatement.setInt(3, entity.getId());

            preparedStatement.executeUpdate();
        } catch (SQLException e) {
            logger.error(e);
            System.out.println("SQLException: " + e.getMessage());
        }
        logger.traceExit();
    }
}